### PR TITLE
Chat microservice

### PR DESCRIPTION
**Pull request by SashkoMolodec**

### CODE REVIEW

File: chat/pom.xml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <version>${junit-jupiter.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.thymeleaf</groupId>
+            <artifactId>thymeleaf-spring5</artifactId>
+            <version>3.0.11.RELEASE</version>
```
Comment (vterlyha) [inline chat/pom.xml:59]: Move version to properties section

File: chat/pom.xml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        </dependency>
+        <dependency>
+            <groupId>javax.persistence</groupId>
+            <artifactId>javax.persistence-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.modelmapper</groupId>
+            <artifactId>modelmapper</artifactId>
+            <version>2.3.5</version>
```
Comment (vterlyha) [inline chat/pom.xml:68]: Same here

File: chat/pom.xml [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.webjars</groupId>
+            <artifactId>jquery</artifactId>
+            <version>3.1.1-1</version>
+        </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt</artifactId>
+            <version>0.9.1</version>
```
Comment (vterlyha) [inline chat/pom.xml:119]: And here

File: chat/src/main/java/greencity/repository/ChatRoomRepo.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+@Repository
+public interface ChatRoomRepo extends JpaRepository<ChatRoom, Long>,
+    JpaSpecificationExecutor<ChatRoom> {
+    /**
+     * Method to find all {@link ChatRoom}'s by {@link Participant}/{@code User} id.
+     *
+     * @param participant {@link Participant} id.
+     * @return list of {@link ChatRoom} instances.
+     */
+    @Query(value = "SELECT dr FROM ChatRoom dr"
```
Comment (vterlyha) [inline chat/src/main/java/greencity/repository/ChatRoomRepo.java:23]: Why abbreviation of ChatRoom is dr ?

File: chat/src/main/java/greencity/service/impl/ChatMessageServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            new TypeToken<List<ChatRoomDto>>() {
+            }.getType());
+    }
+
+    @Override
+    public void processMessage(ChatMessageDto chatMessageDto) {
+        ChatMessage message = modelMapper.map(chatMessageDto, ChatMessage.class);
+        chatMessageRepo.save(message);
+        messagingTemplate.convertAndSend(
+            "/room/" + chatMessageDto.getRoomId() + "/queue/messages", chatMessageDto);
```
Comment (vterlyha) [inline chat/src/main/java/greencity/service/impl/ChatMessageServiceImpl.java:44]: Move strings to constants

File: chat/src/main/java/greencity/service/impl/ChatRoomServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        if (chatRoom.isEmpty()) {
+            toReturn = chatRoomRepo.save(
+                ChatRoom.builder()
+                    .name("chatName")
+                    .participants(participants)
+                    .type(ChatType.PRIVATE)
+                    .build());
+        } else {
+            toReturn = chatRoom.get(0);
+        }
```
Comment (vterlyha) [inline chat/src/main/java/greencity/service/impl/ChatRoomServiceImpl.java:83]: Move to private method

File: core/src/main/java/greencity/controller/RedirectChatController.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.servlet.view.RedirectView;
+
+@Controller
+@AllArgsConstructor
+@RequestMapping("/chat")
+public class RedirectChatController {
+    UserService userService;
```
Comment (vterlyha) [inline core/src/main/java/greencity/controller/RedirectChatController.java:15]: Add private to field

File: chat/pom.xml (hunk 1 above)
Comment (airetOK) [reply chat/pom.xml]: done

File: chat/pom.xml (hunk 2 above)
Comment (airetOK) [reply chat/pom.xml]: done

File: chat/pom.xml (hunk 3 above)
Comment (airetOK) [reply chat/pom.xml]: done

File: chat/src/main/java/greencity/repository/ChatRoomRepo.java (hunk 4 above)
Comment (airetOK) [reply chat/src/main/java/greencity/repository/ChatRoomRepo.java]: done

File: chat/src/main/java/greencity/service/impl/ChatMessageServiceImpl.java (hunk 5 above)
Comment (airetOK) [reply chat/src/main/java/greencity/service/impl/ChatMessageServiceImpl.java]: done

File: core/src/main/java/greencity/controller/RedirectChatController.java (hunk 7 above)
Comment (airetOK) [reply core/src/main/java/greencity/controller/RedirectChatController.java]: done

**Review by forestik — APPROVED** [review]

**Review by Skajl-dev — APPROVED** [review]

**Review by terrytarik — APPROVED** [review]

**Pull request merged**